assembl/indexing/changes.py

```diff
@@ -0,0 +1,153 @@
+import logging
+import threading
+import os
+
+from transaction.interfaces import ISavepointDataManager, IDataManagerSavepoint
```

Comment:
The file changes.py is actually specific to assembl project because of the imports.
I prefer to keep all indexing related code in the indexing package if you don't mind.
Some files in lib folder contain actually domain-specific code, like lib/frontend_urls.py, so there is not really a clear separation here.

You can read http://transaction.readthedocs.io/en/latest/ to know more about data (resource) managers participating in a transaction. We have here two data managers in the transaction, sqlachelmy data manager from zope.sqlachelmy (see venv/lib/python2.7/site-packages/zope/sqlalchemy/datamanager.py), and here elasticsearch I wrote.
We use an elasticsearch data manager here to queue all changes and only put the changes to elasticsearch if postgres successfully committed the transaction.

When a commit occurs, the data managers are first sorted by their sortKey, sqlalchemy is first (key is "~sqlalchemy..."), elasticsearch is second (key is "ZZZZZ...").
Here is the workflow:
- sqlalchemy tpc_begin() does session.flush()
- elasticsearch (changes.py) tpc_begin() does nothing
- sqlalchemy commit() expires objects
- elasticsearch commit() does nothing
- sqlalchemy tpc_vote() commits to postgres. If an exception occurs, it will call tpc_abort on all resource managers, so clearing pending changes for elasticsearch data manager. 
- elasticsearch tpc_vote() does nothing
- sqlalchemy tpc_finish() does nothing
- elasticsearch tpc_finish() write to elasticsearch. If an exception occurs, tpc_abort will be called on all resources managers, but here the changes are already committed to postgres, so it actually doesn't really do anything. But you get a 500 http code.

For the savepoint stuff, you can actually at any moment in your code use `savepoint = transaction.savepoint()` and if there is an exception, rollback with `savepoint.rollback()` to return the transaction to this saved state. For postgres, this is implemented with nested transaction if supported.

I'll put those comments in the changes.py file.